src/Microsoft.DotNet.Tools.Builder/Program.cs

```diff
@@ -0,0 +1,171 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+
+using Microsoft.Dnx.Runtime.Common.CommandLine;
+using Microsoft.DotNet.Cli.Utils;
+using Microsoft.DotNet.ProjectModel;
+using Microsoft.DotNet.Tools.Compiler;
+using NuGet.Frameworks;
+
+
+namespace Microsoft.DotNet.Tools.Build
+{
+    public class Program
+    {
+        public static int Main(string[] args)
+        {
+            DebugHelper.HandleDebugSwitch(ref args);
+
+            try
+            {
+                var app = new CompilerCommandApp("dotnet build", ".NET Builder", "Builder for the .NET Platform. It performs incremental compilation if it's safe to do so. Otherwise it delegates to dotnet-compile which performs non-incremental compilation");
+                return app.Execute(OnExecute, args);
+            }
+            catch (Exception ex)
+            {
+#if DEBUG
+                Console.Error.WriteLine(ex);
+#else
+                Console.Error.WriteLine(ex.Message);
+#endif
+                return 1;
+            }
+        }
+
+        private static bool OnExecute(List<ProjectContext> contexts, string configValue, string outputValue, string intermediateValue, bool noHost,
+            bool isNative, string archValue, string ilcArgsValue, string ilcPathValue, string ilcSdkPathValue,
+            string appDepSdkPath, bool isCppMode)
+        {
+            foreach (var context in contexts)
+            {
+                // Set up Output Paths
+                string outputPath = context.GetOutputPath(configValue, outputValue);
+                string intermediateOutputPath = context.GetIntermediateOutputPath(configValue, intermediateValue, outputValue);
+
+                Directory.CreateDirectory(outputPath);
+                Directory.CreateDirectory(intermediateOutputPath);
+
+                //compile dependencies
+                if (!CompileDependencies(context, configValue, outputPath, intermediateOutputPath, noHost))
+                {
+                    return false;
+                }
+                
+                //compile project
+                if (!CompileProject(context, configValue, outputPath, intermediateOutputPath, noHost, isNative,
```

Comment:
Can you please help elaborate on why do we compile the dependencies of the project and then compile project again from the root (that contains the dependencies compiled earlier)?